R/RandMat.R

```diff
@@ -1,210 +1,508 @@
+#'
+#' @return A random matrix to use in running \code{\link{RerF}}.
 #'
 #' @export
 #'
-#' @importFrom RcppZiggurat zrnorm
-#' @importFrom stats rpois runif
-
-RandMat <-
-function(mat.options) {
-  p <- mat.options[[1L]] # number of dimensions
-
-  d <- mat.options[[2L]] # this determines the number of columns in the projection matrix.
-  method <- mat.options[[3L]] # defines the distribution of the random projection matrix
-  #Create the random matrix, a sparse matrix of 1's, -1's, and 0's.
-  if (method == "binary") {
-    rho <- mat.options[[4L]]
-    prob <- mat.options[[5L]]
-    nnzs <- round(p*d*rho)
-    ind <- sort(sample.int((p*d), nnzs, replace = FALSE))
-    random.matrix <- cbind(((ind - 1L) %% p) + 1L, floor((ind - 1L) / p) + 1L,
-                           sample(c(1L, -1L), nnzs, replace = TRUE, prob = c(prob, 1 - prob)))
-  } else if (method == "continuous") {
-    rho <- mat.options[[4L]]
-    nnzs <- round(p*d*rho)
-    ind <- sort(sample.int((p*d), nnzs, replace = FALSE))
-    random.matrix <- cbind(((ind - 1L) %% p) + 1L, floor((ind - 1L) / p) + 1L,
-                           zrnorm(nnzs))
-  } else if (method == "rf") {
-    random.matrix <- cbind(sample.int(p, d, replace = FALSE), 1:d, rep(1L, d))
-  } else if (method == "poisson") {
-    lambda <- mat.options[[4L]]
-    if (lambda <= 0) { 
-      stop(" Wrong parameter for Poisson, make sure lambda > 0.") 
-    }
-    go <- TRUE
-    while (go) {
-      nnzPerCol <- stats::rpois(d, lambda)
-      go <- !any(nnzPerCol)
+#' @examples
+#'
+#' p <- 8
+#' d <- 3
+#' rho <- 0.25
+#' prob <- 0.5
+#' set.seed(4)
+#' (a <- RandMatBinary(p, d, rho, prob))
+#'
+
+RandMatBinary <- function(p, d, rho, prob, catMap = NULL) {
+  nnzs <- round(p * d * rho)
+  ind <- sort(sample.int((p * d), nnzs, replace = FALSE))
+  
+  ## Determine if categorical variables need to be taken into
+  ## consideration
+  if (is.null(catMap)) {
+    randomMatrix <- cbind(((ind - 1L)%%p) + 1L, floor((ind - 1L)/p) + 
+      1L, sample(c(1L, -1L), nnzs, replace = TRUE, prob = c(prob, 
+      1 - prob)))
+  } else {
+    pnum <- catMap[[1L]][1L] - 1L
+    rw <- ((ind - 1L)%%p) + 1L
+    isCat <- rw > pnum
+    for (j in (pnum + 1L):p) {
+      isj <- rw == j
+      rw[isj] <- sample(catMap[[j - pnum]], length(rw[isj]), replace = TRUE)
     }
-    nnzPerCol[nnzPerCol > p] <- p
-    nnz <- sum(nnzPerCol)
-    nz.rows <- integer(nnz)
-    nz.cols <- integer(nnz)
-    start.idx <- 1L
-    for (i in seq.int(d)) {
-      if (nnzPerCol[i] != 0L) {
-        end.idx <- start.idx + nnzPerCol[i] - 1L
-        nz.rows[start.idx:end.idx] <- sample.int(p, nnzPerCol[i], replace = FALSE)
-        nz.cols[start.idx:end.idx] <- i
-        start.idx <- end.idx + 1L
-      }
+    randomMatrix <- cbind(rw, floor((ind - 1L)/p) + 1L, sample(c(1L, 
+      -1L), nnzs, replace = TRUE, prob = c(prob, 1 - prob)), deparse.level = 0)
+  }
+  return(randomMatrix)
+}  ## END RandMatBinary
```

Comment:
can we remove these if they aren't needed anymore?

Reply:
Done.